pkg/expansion/system.go

```diff
@@ -176,10 +176,27 @@ func expandResource(obj *unstructured.Unstructured, ns *corev1.Namespace, templa
 	resource.SetUnstructuredContent(src)
 	resource.SetGroupVersionKind(resultantGVK)
 	resource.SetNamespace(ns.Name)
+	resource.SetName(mockNameForResource(obj, resultantGVK))
 
 	return resource, nil
 }
 
+// mockNameForResource returns a mock name for a resultant resource created
+// from expanding `gen`. The name will be of the form:
+// "mock-<resultant kind>-<generator name>". For example, a deployment named
+// `nginx-deployment` will produce a resultant named `mock-pod-nginx-deployment`.
+func mockNameForResource(gen *unstructured.Unstructured, gvk schema.GroupVersionKind) string {
```

Comment:
this may be a bad question but are there scenarios where resources get expanded multiple times? or a situation in which we have the same generated resource w the same `mock` name? if so, would that be a concern? would we want to suffix this w a rand string?

Reply:
Recursive expansion is not supported but there is a request for it so it's might a valid point if implemented.

https://github.com/open-policy-agent/gatekeeper/issues/2511

If it were you'd have Cronjob "cleanup" expanded to "cleanup-mock-job" and again to "cleanup-mock-job-mock-pod". Could either deem that acceptable, strip off the -mock-* suffix before generating the next child mock name resulting in just "cleanup-mock-pod" for the last expansion, or everything after mock is implied recursive expansion so only append each additional mock object "cleanup-mock-job-pod".

Reply:
@acpana The same resource can be expanded multiple times if there are multiple `ExpansionTemplates` that match it. In this case, presumably, the [`GeneratedGVK`](https://github.com/open-policy-agent/gatekeeper/blob/master/apis/expansion/v1alpha1/expansion_template_types.go#L42) would be different, which would make the resultant resources' names different.

It is technically possible for 2 `ExpansionTemplates` to match the same Kind and produce the same `GeneratedGVK`, but I don't think the 2 resultant resources with the same name would present any technical issues, as the expanded resources are ephemeral.

Reply:
@ctrought Great point, thanks for calling this out.

I think appending "mock-*" for each recursive expand is redundant. I like your last suggestion to only specify `mock` after the first recursion (i.e `cleanup-mock-job-pod`). It looks like the resource name limit is [253 characters](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names), but I would think this should be sufficient. 

@maxsmythe @acpana thoughts?

Reply:
For glob-matching purposes I would not prepend `mock-*` (suffix would work).

I think it's slightly simpler to just copy the name of the parent resource, but the difference between that and a `-mock` suffix (including preventing recursion) is close enough that I don't mind either way.

One extra edge case -- what if the parent resource name ends in `-mock`? How would the anti-recursion logic deal with that? I presume by not appending an extra `-mock`, which means users may or may not successfully re-derive the parent object name by stripping the suffix. That edge cases keep showing up is why I tend to lean toward just copying the parent name.